File: ios/HybridCactusCrypto.swift
import CryptoKit
import NitroModules

class HybridCactusCrypto: HybridCactusCryptoSpec {
  func uuidv5(namespaceUuid: String, name: String) throws -> Promise<String> {
    return Promise.async {
      guard let ns = UUID(uuidString: namespaceUuid) else {
        throw RuntimeError.error(withMessage: "Invalid namespace UUID")
      }

      var nsBytes = ns.uuid
      var data = withUnsafeBytes(of: &nsBytes) { Data($0) }
      data.append(Data(name.utf8))

      let digest = Insecure.SHA1.hash(data: data)
      var bytes = Data(digest.prefix(16))
      bytes.withUnsafeMutableBytes { raw in
        let b = raw.bindMemory(to: UInt8.self)
        b[6] = (b[6] & 0x0F) | 0x50
        b[8] = (b[8] & 0x3F) | 0x80
      }

      let tuple = bytes.withUnsafeBytes { ptr -> uuid_t in
        let b = ptr.bindMemory(to: UInt8.self)
        return uuid_t(
          b[0], b[1], b[2], b[3],
          b[4], b[5], b[6], b[7],
          b[8], b[9], b[10], b[11],
          b[12], b[13], b[14], b[15])
      }

      let uuid = UUID(uuid: tuple)

      return uuid.uuidString.lowercased()
    }
  }
}

File: android/src/main/java/com/margelo/nitro/cactus/HybridCactusCrypto.kt
package com.margelo.nitro.cactus

import com.margelo.nitro.core.Promise
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.Locale
import java.util.UUID

class HybridCactusCrypto : HybridCactusCryptoSpec() {
  override fun uuidv5(
    namespaceUuid: String,
    name: String,
  ): Promise<String> =
    Promise.async {
      val nsUuid =
        try {
          UUID.fromString(namespaceUuid)
        } catch (e: IllegalArgumentException) {
          throw IllegalArgumentException("Invalid namespace UUID")
        }

      val nsBytes =
        ByteBuffer
          .allocate(16)
          .apply {
            putLong(nsUuid.mostSignificantBits)
            putLong(nsUuid.leastSignificantBits)
          }.array()
      val nameBytes = name.toByteArray(Charsets.UTF_8)

      val sha1 =
        MessageDigest
          .getInstance("SHA-1")
          .apply { update(nsBytes) }
          .digest(nameBytes)
      val uuidBytes = sha1.copyOfRange(0, 16)
      uuidBytes[6] = (uuidBytes[6].toInt() and 0x0F or 0x50).toByte()
      uuidBytes[8] = (uuidBytes[8].toInt() and 0x3F or 0x80).toByte()

      val bb = ByteBuffer.wrap(uuidBytes)

      val uuid = UUID(bb.long, bb.long)

      uuid.toString().lowercase(Locale.ROOT)
    }
}
